### ISSUE TITLE
Prestodb version 0.237 has lower performance compared with version 0.215

### ISSUE THREAD

**Issue by Xiangakun** [prestodb/presto#14978]
When I am going to upgrade prestodb version from 0.215 to 0.237, I try the TPC-H test, and the test result show that, the higher version has poor performance than the lower version, especially when the concorrent sql num is growing, the gap will be more obvious.

Test data:  TPC-H test data 100 G
Sqls:  totally 22 TPC-H sqls
Test method:  for each sql，firstly run  10 times (one by one), and then 10,20,30,40,50,60 concurrent sqls at the same time
Test result：calculate the average time and median time，for most sqls (except for: tpch-sql1,sql2,sql6,sql15,sql19)，0.237 has poor performance than 0.215

Test env:   1 coordinator，25 workers in k8s cloud env
Configurations:  Cpus 30 and Mem 204,800(MB), JDK8
JVM config: 
   -Xmx163840M
  -XX:MaxDirectMemorySize=20480M
  -Xms163840M
  -XX:ReservedCodeCacheSize=2048M
  -Xss8M
  -XX:G1ReservePercent=15
  -XX:+UseG1GC
  -XX:MaxGCPauseMillis=200

config.properties config:
   query.max-memory=4096GB
   query.max-memory-per-node=64GB
   query.initial-hash-partitions=50
   query.max-total-memory-per-node=72.0GB
   task.max-worker-threads=120
   query.min-expire-age=5m
   query.max-history=80
   query.max-run-time=2h
   http-server.max-request-header-size=5MB
   exchange.min-error-duration=2m
   task.http-timeout-threads=10
   experimental.reserved-pool-enabled=false
   memory.heap-headroom-per-node=20GB
   query.low-memory-killer.policy=total-reservation-on-blocked-nodes

Note: Sqls run at the same time within both cluster 0.237 and 0.215, that is, we can ignore the effect from the hadoop env, cpu, io etc.

**Comment by wenleix** [comment]
@mbasmanova ,  @yingsu00 : Wondering if you have any thoughts into this ? Thanks!

**Comment by yingsu00** [comment]
@wenleix were you using Aria? Was it only concurrent runs are slower? Can you share the query ids for these runs?

**Comment by Xiangakun** [comment]
Following is the file which includes my test result and the test sql, thx~

https://github.com/Xiangakun/link/blob/master/presto/tpch-presto-215vs237-concurrent-perfance-test.xlsx

**Comment by Xiangakun** [comment]
For some cases, it's easier to met the error **"The node may have crashed or be under too much load. This is probably a transient issue, so please retry your query in a few minutes."** in version 0.237. In addition, I compared with the time each stage/task/operator took, for some cases, it seems the table scan operator takes more time, and, some cases, it seems to be the exchange operator. I am wondering if it's caused by the communication between workers, which has lower performace.

**Comment by mbasmanova** [comment]
@Xiangakun To find out where the slowdown is coming from, please, collect CPU profiles using async profiler: https://github.com/jvm-profiling-tools/async-profiler

**Comment by Xiangakun** [comment]
@mbasmanova Thanks for your advice, I tried test case 22.sql with 50 concurrency，the test result is : average time spent on 0.215 was about 50~51s, and average time spent on 0.237 was about 1.1m. As you advised, I collected the CPU profiles during the running time. Please refer to the svg files under the link fold (I randomly collect the cpu profiles of only one worker): https://github.com/Xiangakun/link/tree/master/presto    
It shows the most cpu time spend are: memory set, Hash operation and GC, and it seems there aren't much difference between 0.215  and 0.237.  However it did take more cpu time when doing memory set in 0.237, I am not sure it's the cause or just the result.

**Comment by mbasmanova** [comment]
@Xiangakun Is it possible to reproduce this on a single query?

**Comment by Xiangakun** [comment]
@mbasmanova   Yes, I think so. However for a single query, the gap is not obvious, since in my env,  each single query finish within few seconds. I am wondering if it's possible for you to try the tests by yourself in your local, and make sure the problem did exists. Thanks in advance!

**Comment by mbasmanova** [comment]
@Xiangakun I don't have this kind of setup. TPC-H queries doesn't seem to be representative of the real world workflows, hence, we don't invest in testing them.

**Comment by Xiangakun** [comment]
@mbasmanova If so, do you have any suggestions about how to test and then find out whether the new prestodb version has any performance improved？

**Comment by yingsu00** [comment]
@Xiangakun We don't have TPCH tracking runs yet. We will have it by the end of this year. Is this regression really important to you? If so can you pick a most regressed query, run it several times and share with us the performance data after excluding the first run:
-  query elapsed times and variation
- query CPU times and variation
- If CPU time regressed, take aync profiler profile for both version

When doing the above perf testing, make sure all queries succeeded and only take the warm runs.

From the profile you shared, the network stack's (jetty) cost increased and the driver's cost doesn't seem to have regressed. It looks like the problem is not in the Presto execution engine.
